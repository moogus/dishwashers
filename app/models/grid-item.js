import { shape, string } from 'prop-types';

export const gridItemProps = shape({
  productId: string,
  price: string,
  title: string,
  image: string,
});

export const buildModel = (gridItem = {}) => {
  const {
    productId,
    price = {},
    title,
    image,
  } = gridItem;

  const { now } = price;

  return {
    productId,
    price: `Â£${now}`,
    title,
    image,
  };
};
