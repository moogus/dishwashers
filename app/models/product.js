export const buildModel = (product = {}) => {
  const {
    title,
    media = {},
    price = {},
    details = {},
    displaySpecialOffer,
    additionalServices,
    code,
  } = product;

  const { now = '' } = price;
  const { productInformation } = details;
  const { images = {} } = media;
  const { altText = '', urls = [] } = images;

  return {
    title,
    media: { altText, urls },
    price: `Â£${now}`,
    productInformation,
    displaySpecialOffer,
    additionalServices,
    code,
  };
};
