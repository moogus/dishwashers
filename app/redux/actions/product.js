export const PRODUCT_LOADING = 'PRODUCT_LOADING';
export const PRODUCT_LOADED = 'PRODUCT_LOADED';
export const PRODUCT_ERROR = 'PRODUCT_ERROR';

export const loadProduct = (productId) => async (dispatch, _, config) => {
  dispatch({ type: PRODUCT_LOADING });

  try {
    const response = await fetch(`${config.productApi}${productId}`);
    const product = await response.json();

    dispatch({ type: PRODUCT_LOADED, product });
  } catch (err) {
    dispatch({ type: PRODUCT_ERROR });
  }
};
