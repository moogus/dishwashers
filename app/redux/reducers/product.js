import {
  PRODUCT_LOADING,
  PRODUCT_LOADED,
  PRODUCT_ERROR,
} from 'actions/product';

const intialState = { product: {}, loading: true };

export default (state = intialState, action) => {
  switch (action.type) {
    case PRODUCT_LOADING: {
      return {
        ...state,
        loading: true,
      };
    }

    case PRODUCT_LOADED: {
      const { product = {} } = action;

      return {
        ...state,
        product: {
          ...product,
        },
        loading: false,
      };
    }

    case PRODUCT_ERROR: {
      return {
        ...state,
        loading: false,
        error: true,
      };
    }

    default:
      return state;
  }
};
