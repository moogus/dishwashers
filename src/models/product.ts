import { ApiProduct, Product } from "../types"

export const buildModel = (apiProduct:ApiProduct):Product => {
  const {
    title,
    media,
    price,
    details,
    displaySpecialOffer,
    additionalServices,
    code,
  } = apiProduct;

  const { now } = price;
  const { productInformation } = details;
  const { images } = media;
  const { altText, urls } = images;

  return {
    title,
    media: { altText, urls },
    price: `Â£${now}`,
    productInformation,
    displaySpecialOffer,
    additionalServices,
    code,
  };
};
