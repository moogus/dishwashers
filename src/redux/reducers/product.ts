import {
  PRODUCT_LOADING,
  PRODUCT_LOADED,
  PRODUCT_ERROR,
} from "../actions/product";

import { ApiProduct } from "../../types";
import { ProductsState } from "../../interfaces" 


export const intialState:ProductsState = { 
  product: {} as ApiProduct, 
  loading: true, 
  error: false
};

export default (state = intialState, action: { type?: string; product?: ApiProduct; }):ProductsState  => {
  switch (action.type) {
    case PRODUCT_LOADING: {
      return {
        ...state,
        loading: true,
      };
    }

    case PRODUCT_LOADED: {
      const { product } = action;

      return {
        ...state,
        product,
        loading: false,
      };
    }

    case PRODUCT_ERROR: {
      return {
        ...state,
        loading: false,
        error: true,
      };
    }

    default:
      return state;
  }
};
