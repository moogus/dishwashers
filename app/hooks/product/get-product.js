import { useSelector } from 'react-redux';
import { buildModel } from 'models/product';

export default () => {
  const { product } = useSelector((state) => state.product);
  return buildModel(product);
};
