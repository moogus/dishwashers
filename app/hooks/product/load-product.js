import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { loadProduct } from 'actions/product';

export default (productId) => {
  const dispatch = useDispatch();
  const { loading, error } = useSelector((state) => state.product);

  useEffect(() => {
    dispatch(loadProduct(productId));
    // eslint-disable-next-line
  }, []);

  return { loading, error };
};
