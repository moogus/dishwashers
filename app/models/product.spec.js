import { buildModel } from 'models/product';

describe('model is built from data', () => {
  test('empty model is build as expected', () => {
    expect(buildModel()).toMatchSnapshot();
  });

  test('hydrated model is build as expected', () => {
    expect(buildModel({
      title: 'come a long way',
      media: {
        images: {
          alText: 'foo',
          urls: ['1'],
        },
      },
      price: { now: 'right here' },
      details: {},
      displaySpecialOffer: '',
      additionalServices: {},
      code: '1234',
    })).toMatchSnapshot();
  });
});
