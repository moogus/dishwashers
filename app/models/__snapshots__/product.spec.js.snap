// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model is built from data empty model is build as expected 1`] = `
Object {
  "additionalServices": undefined,
  "code": undefined,
  "displaySpecialOffer": undefined,
  "media": Object {
    "altText": "",
    "urls": Array [],
  },
  "price": "£",
  "productInformation": undefined,
  "title": undefined,
}
`;

exports[`model is built from data hydrated model is build as expected 1`] = `
Object {
  "additionalServices": Object {},
  "code": "1234",
  "displaySpecialOffer": "",
  "media": Object {
    "altText": "",
    "urls": Array [
      "1",
    ],
  },
  "price": "£right here",
  "productInformation": undefined,
  "title": "come a long way",
}
`;
